File: apps/mobile/ios/Sources/FlagshipUI/ViewModels/ServicesListViewModel.swift
import Foundation
import Observation
import FlagshipAPI

@Observable
@MainActor
public final class ServicesListViewModel {
    public private(set) var state: LoadingState<[FlagshipAPI.AppSummary]> = .idle
    /// V2 — per-service links cache. Populated by `loadLinks` after the
    /// daemon's apps-list returns. Keyed by serviceId. Absent means
    /// either still loading or .com hasn't been reached; the row
    /// falls back to the daemon-provided urlLabel.
    public private(set) var linksByServiceId: [String: AppLinksResponse] = [:]
    public var searchQuery: String = ""
    /// V7 — server filter. `nil` = "All servers" (the default). When
    /// set to a specific podId, the services list is filtered to services
    /// running on that pod. Today's daemon AppsListResponse doesn't
    /// yet carry a `installedOn` field per service, so we approximate by
    /// matching the pod name against the canonical FQDN the daemon
    /// surfaces (e.g. `notes.home.alice.flagship.services` belongs
    /// to "home"). Once the daemon emits the membership list this
    /// match becomes exact.
    public var serverFilter: String? = nil
    /// V7 — pods the user has on file. The Services tab passes the
    /// current AppState.pods snapshot in so the dropdown can list
    /// each server by name. Default empty so the dropdown is hidden
    /// when the user has no pods yet.
    public var availablePods: [(podId: String, name: String)] = []

    private let client: any ScreensClient
    private let server: (any FlagshipServerClient)?
    private let username: () -> String?

    public init(
        client: any ScreensClient,
        server: (any FlagshipServerClient)? = nil,
        username: @escaping () -> String? = { nil }
    ) {
        self.client = client
        self.server = server
        self.username = username
    }

    public var filteredApps: [FlagshipAPI.AppSummary] {
        guard let apps = state.value else { return [] }
        var out = apps
        if let podId = serverFilter,
           let podName = availablePods.first(where: { $0.podId == podId })?.name {
            let needle = podName.lowercased()
            // Match services whose canonical URL contains the pod name as
            // a subdomain segment. The daemon's AppSummary.url has
            // the shape `https://<label>.<server>.<user>.flagship.services`,
            // so checking the lowercased URL is a stable approximation
            // until a real installedOn field lands.
            out = out.filter { $0.url.lowercased().contains(".\(needle).") }
        }
        if searchQuery.isEmpty { return out }
        let q = searchQuery.lowercased()
        return out.filter {
            $0.serviceId.lowercased().contains(q)
                || $0.slug.lowercased().contains(q)
                || ($0.summary?.lowercased().contains(q) ?? false)
        }
    }

    public func load() async {
        state = .loading
        do {
            let resp = try await client.appsList()
            state = .loaded(resp.apps)
            // Kick off link fetches in parallel — surfaced as they
            // arrive so the list paints fast and the URLs fill in.
            await loadLinks(for: resp.apps)
        } catch {
            state = .failed(error.localizedDescription)
        }
    }

    /// V2 — fan-out fetch of /api/users/:u/apps/:serviceId/links per service.
    /// Tolerates per-service failure (e.g. .com momentarily down for one
    /// row) without nuking the whole list. Updates `linksByServiceId`
    /// as each result lands.
    private func loadLinks(for apps: [FlagshipAPI.AppSummary]) async {
        guard let server, let user = username(), !user.isEmpty else { return }
        await withTaskGroup(of: (String, AppLinksResponse?).self) { group in
            for app in apps {
                group.addTask { [serviceId = app.serviceId] in
                    do {
                        let r = try await server.getAppLinks(username: user, serviceId: serviceId)
                        return (serviceId, r)
                    } catch {
                        return (serviceId, nil)
                    }
                }
            }
            for await (serviceId, links) in group {
                if let links { linksByServiceId[serviceId] = links }
            }
        }
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/viewmodels/ServicesListViewModel.kt
// Kotlin mirror of FlagshipUI/ViewModels/AppsListViewModel.swift.
//
// Loads the daemon's paired-session apps-list, then fans out a
// per-service /api/users/:u/apps/:serviceId/links fetch (#20). The list
// paints as soon as appsList() returns; each row's voi.ci short URL,
// canonical FQDN and bound custom domain fill in as the /links
// results land. Per-service failure is tolerated (one row's .com blip
// must not nuke the whole list) — exactly the iOS behavior.
//
// When the contract changes, update this file AND
// apps/mobile/ios/Sources/FlagshipUI/ViewModels/AppsListViewModel.swift
// in the same commit.

package com.flagshipserver.app.viewmodels

import com.flagshipserver.app.api.AppLinksResponse
import com.flagshipserver.app.api.AppSummary
import com.flagshipserver.app.api.FlagshipServerClient
import com.flagshipserver.app.api.ScreensClient
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class ServicesListViewModel(
    private val client: ScreensClient,
    private val server: FlagshipServerClient?,
    private val username: () -> String?,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.IO),
) {
    private val _state = MutableStateFlow<LoadingState<List<AppSummary>>>(LoadingState.Idle)
    val state: StateFlow<LoadingState<List<AppSummary>>> = _state.asStateFlow()

    /** Per-service links cache, keyed by serviceId. Absent means still
     *  loading or .com unreachable for that row; the row falls back
     *  to the daemon-provided URL. Mirrors iOS `linksByServiceId`. */
    private val _linksByServiceId = MutableStateFlow<Map<String, AppLinksResponse>>(emptyMap())
    val linksByServiceId: StateFlow<Map<String, AppLinksResponse>> = _linksByServiceId.asStateFlow()

    fun load() = scope.launch {
        _state.value = LoadingState.Loading
        try {
            val resp = client.appsList()
            _state.value = LoadingState.Loaded(resp.apps)
            // Kick off the link fan-out — rows paint immediately and
            // the URLs fill in as each result arrives.
            loadLinks(resp.apps)
        } catch (t: Throwable) {
            _state.value = LoadingState.Failed(t.message ?: "couldn't load services")
        }
    }

    /** V2 — fan-out fetch of /api/users/:u/apps/:serviceId/links per
     *  service. Tolerates per-service failure without nuking the list;
     *  updates `linksByServiceId` as each result lands. */
    private suspend fun loadLinks(apps: List<AppSummary>) = coroutineScope {
        val srv = server ?: return@coroutineScope
        val user = username()?.takeIf { it.isNotEmpty() } ?: return@coroutineScope
        val deferred = apps.map { app ->
            async {
                try {
                    app.serviceId to srv.getAppLinks(user, app.serviceId)
                } catch (t: Throwable) {
                    app.serviceId to null
                }
            }
        }
        for (d in deferred) {
            val (serviceId, links) = d.await()
            if (links != null) {
                _linksByServiceId.value = _linksByServiceId.value + (serviceId to links)
            }
        }
    }
}
